src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -179,12 +192,14 @@ Status CoreWorkerPlasmaStoreProvider::Get(
     }
 
     size_t previous_size = remaining.size();
-    // TODO: For direct calls, use NotifyDirectCallTaskBlocked/Unblocked calls
-    // for missing objects instead of going through the normal fetch-and-get
-    // codepath.
-    RAY_RETURN_NOT_OK(FetchAndGetFromPlasmaStore(remaining, batch_ids, batch_timeout,
-                                                 /*fetch_only=*/false, task_id, results,
-                                                 got_exception));
+    // This is a separate IPC from the FetchAndGet in direct call mode.
+    if (ctx.CurrentTaskIsDirectCall() && ctx.ShouldReleaseResourcesOnBlockingCalls()) {
+      RAY_RETURN_NOT_OK(raylet_client_->NotifyDirectCallTaskBlocked());
+    }
+    RAY_RETURN_NOT_OK(
+        FetchAndGetFromPlasmaStore(remaining, batch_ids, batch_timeout,
+                                   /*fetch_only=*/ctx.CurrentTaskIsDirectCall(),
```

Comment:
Hmm actually won't this break the eviction errors? Because right now, we rely on the call to the task_dependency_manager_ [here](https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L1953) to cause the Get to timeout and store an UnreconstructableError.

Reply:
Hm I see, so IIUC fetch won't store errors?

That sounds unfortunate, so do we need to restore the mark_worker_blocked flag and pass it all the way into HandleTaskBlocked (maybe rename HandleTaskBlocked to something more sensible?)

Reply:
Yeah, sorry I didn't realize this earlier.

Another option is to duplicate the task dependency manager calls in the HandleDirectCallTaskBlocked/Unblocked handlers. It's not clear to me which one would be cleaner (aka less messy).

Reply:
I ended up doing the former. It is quite messy now, but I guess once we delete the old path we can remove the code guarded by the flag entirely and it will be clean again.